File: 39.cpp
struct Solution {
    void gen(int idx, int remain, vector<int> &prefix, vector<vector<int>> &res, vector<int> &cand){
        if (remain == 0){
            res.push_back(prefix);
            return;
        }
        if (idx == cand.size() || cand[idx] > remain){
            return;
        }
        this->gen(idx + 1, remain, prefix, res, cand);
        int x = cand[idx];
        int r = remain / x;
        for (int i = 1; i <= r; ++i){
            prefix.push_back(x);
            this->gen(idx + 1, remain - x * i, prefix, res, cand);
        }
        while (r-- > 0){
            prefix.pop_back();
        }
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> res;
        vector<int> prefix;
        this->gen(0, target, prefix, res, candidates);
        return res;
    }
};

File: sol/39.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        res = []
        n = len(candidates)
        def gen(idx, remain, prefix):
            if remain == 0:
                res.append(prefix.copy())
                return
            if idx == n or candidates[idx] > remain:
                return
            gen(idx + 1, remain, prefix)
            x = candidates[idx]
            r = remain // x
            for i in range(1, r + 1):
                prefix.append(x)
                gen(idx + 1, remain - x * i, prefix)
            for _ in range(r):
                prefix.pop()
        gen(0, target, [])
        return res
